### PR TITLE
Flip "opaque_water" setting to "translucent_liquids"

### PR DESCRIPTION
**Pull request by Xeno333**
Closes Issue #11330 
by fliping "opaque_water" setting to "translucent_liquids"

### CODE REVIEW

File: src/nodedef.cpp
Comment (appgurueu) [inline src/nodedef.cpp:828]: You don't need the parentheses here; same below.

File: minetest.conf.example
Comment (appgurueu) [inline minetest.conf.example:170]: You need not edit this file (`minetest.conf.example`); it is autogenerated from `builtin/settingtypes.txt` when releasing.

**Review by appgurueu — COMMENTED** [review]
This looks like a proper implementation of the "first approach" outlined in that issue (renaming the setting, flipping its meaning).

The only problem I see with this is that it effectively "breaks" the old setting. Are we fine with that? (And if not, is there an acceptably easy way to add "backwards compatibility"?)

**Comment by Xeno333** [comment]
> This looks like a proper implementation of the "first approach" outlined in that issue (renaming the setting, flipping its meaning).
> 
> The only problem I see with this is that it effectively "breaks" the old setting. Are we fine with that? (And if not, is there an acceptably easy way to add "backwards compatibility"?)

Yes I can add a check for the legacy if the new setting is not found.

**Comment by Xeno333** [comment]
I belive I have fixed it am compiling rn to test

**Comment by Xeno333** [comment]
I need to make one more change.

**Comment by Xeno333** [comment]
I am activly working on adding support for legacy. If this is deemed unneeded please tell me.

**Comment by Xeno333** [comment]
I am testing it now. I think I have it ready.

**Comment by Xeno333** [comment]
It does not use the old setting if present and defaults to the being transparent, but if there is no new value set with the new setting. I can not find any simple way to use the old setting, and allow it to be changed to the new setting, perhaps I can reset the new setting with an inverted value of the old setting.

**Comment by Xeno333** [comment]
I think something like:
`
	if (g_settings->exists("opaque_water")) {
		translucent_water              = !(g_settings->getBool("opaque_water"));
		//resave as translucent_water
	}
	translucent_water              = !(g_settings->getBool("translucent_water"));
`

Would be the best way to handel this.

**Comment by Xeno333** [comment]
I havent tried to do a push from terminal lately and messed up the commits:
I closed and reopend here (#14662 ) with fully working code

**Comment by Xeno333** [comment]
OK its working now.

File: src/nodedef.cpp
Comment (Xeno333) [inline src/nodedef.cpp:286]: Add block that handels opaque_water and translucent_water

**Comment by Xeno333** [comment]
Ok it does work, and seems to properly handel the legacy setting. There seems to be no conflict now.

**Comment by Xeno333** [comment]
@appgurueu I have added backwards support and have tested it. It now will respect the old setting until you update it as the new setting.

File: src/nodedef.cpp
Comment (appgurueu) [inline src/nodedef.cpp:287]: This can be simplified to

```c++
if (g_settings->exists("opaque_water"))
    g_settings->setDefault("translucent_water", g_settings->getBool("opaque_water") ? "false" : "true");
translucent_water = g_settings->getBool("translucent_water");
```

However, I'm not very sold on changing the global setting defaults here. Why not instead do the following:

```c++
// Fall back to old setting with inverted meaning if new setting is absent
if (g_settings->exists("opaque_water") && !g_settings->exists("translucent_water")) {
    translucent_water = !g_settings->getBool("opaque_water");
} else {
    translucent_water = g_settings->getBool("translucent_water");
}
```

This would avoid mutating `g_settings`, hence staying true to the name of `readSettings`.

**Review by appgurueu — COMMENTED** [review]
I'm afraid there is a misunderstanding; I did not ask you to make changes, I just wanted to clarify some questions. (This does not mean that I want you to undo the changes.)

I think "migrating" (or at least respecting the old setting) is probably a good solution, though care needs to be taken to implement, and possibly document, this properly, since this will set a precedent for "migrating" settings.

File: src/nodedef.cpp
Comment (Xeno333) [reply src/nodedef.cpp:287]: I tried that origonaly, but it seemed conflicted with the default setting. I can try again wit your code. I do agree that would be the best way though.

Comment (Xeno333) [reply src/nodedef.cpp:287]: Just tested that does NOT work because it conflicts with the defined default so that no matter what `g_settings->exists("translucent_water")` does exist thus will never run.

**Comment by Xeno333** [comment]
> I'm afraid there is a misunderstanding; I did not ask you to make changes, I just wanted to clarify some questions. (This does not mean that I want you to undo the changes.)
> 
> I think "migrating" (or at least respecting the old setting) is probably a good solution, though care needs to be taken to implement, and possibly document, this properly, since this will set a precedent for "migrating" settings.

I understand and understood. I agreed that compatability is needed, thats why i updated it originaly.

**Comment by Xeno333** [comment]
I am for using,
`
	if (g_settings->exists("opaque_water"))
    	g_settings->setDefault("translucent_water", g_settings->getBool("opaque_water") ? "false" : "true");
	translucent_water = g_settings->getBool("translucent_water");
`
Insted of my oringal block. It is much clearer, as I couldnt find a `setDefault` at the time I wrote it + the fact I forgot that `? "false" : "true"` works in C++, as I tend to use plain C where you can't directly compare strings (Objects)

**Comment by Xeno333** [comment]
> I think "migrating" (or at least respecting the old setting) is probably a good solution, though care needs to be taken to implement, and possibly document, this properly, since this will set a precedent for "migrating" settings.

Perhaps there should be a function on called `Migrate::MigrateSettings()` that does this. The file would be Migrate.cpp and it would contain the code for migrating away from old features, this way we can implement whatever change we want as long as we account for it in Migrate.cpp

File: src/nodedef.cpp
Comment (appgurueu) [reply src/nodedef.cpp:287]: My bad, `exists` also checks for defaults. We could use `existsLocal`, but that would only check a single layer.

I think one way to solve it would be to introduce a method `Settings *Settings::getTopLayer(const std::string &setting)` which returns you a pointer to the top layer on which a setting was found, or `nullptr`. This could then be used via `g_settings->getTopLayer("translucent_water") != g_settings->getLayer(SL_DEFAULTS)`.

This should work (untested):
```diff
diff --git a/src/settings.cpp b/src/settings.cpp
index 74d00762d..2eadede52 100644
--- a/src/settings.cpp
+++ b/src/settings.cpp
@@ -657,14 +657,18 @@ bool Settings::getNoiseParamsFromGroup(const std::string &name,
 	return true;
 }
 
-
-bool Settings::exists(const std::string &name) const
+Settings *Settings::getTopLayer(const std::string &name) const
 {
 	if (existsLocal(name))
-		return true;
+		return this;
 	if (auto parent = getParent())
-		return parent->exists(name);
-	return false;
+		return parent->getTopLayer(name);
+	return nullptr;
+}
+
+bool Settings::exists(const std::string &name) const
+{
+	return getTopLayer(name) != nullptr;
 }
 
 
diff --git a/src/settings.h b/src/settings.h
index 692001c59..5d44fa158 100644
--- a/src/settings.h
+++ b/src/settings.h
@@ -174,6 +174,8 @@ class Settings {
 
 	// return all keys used in this object
 	std::vector<std::string> getNames() const;
+	// get the top layer on which a setting exists, or nullptr if it does not exist
+	Settings *getTopLayer(const std::string &name) const;
 	// check if setting exists anywhere in the hierarchy
 	bool exists(const std::string &name) const;
 	// check if setting exists in this object ("locally")
```

**Review by appgurueu — APPROVED** [review]
Works, thanks.

**Comment by Xeno333** [comment]
Sorry, I force pushed again becaus I had wrong email in my git and it didnt show as from me, lol,

File: builtin/settingtypes.txt
Comment (grorp) [inline builtin/settingtypes.txt:268]: "Makes water translucent" isn't true. The setting doesn't make water transparent, but merely allows it to be transparent. If a game has opaque water, it still won't be transparent even if the setting is enabled.

**Review by grorp — COMMENTED** [review]
All liquids are affected by this setting, not just water. Not sure whether this is relevant in practice (is there a translucent liquids other than water in any game for Minetest?), but if we're renaming and migrating the setting anyway, it would make sense to take that into account.

**Comment by Xeno333** [comment]
> All liquids are affected by this setting, not just water. Not sure whether this is relevant in practice (is there a translucent liquid other than water in any game for Minetest?), but if we're renaming and migrating the setting anyway, it would make sense to take that into account.

Would you like me to make it "translucent_liquid" ?

File: builtin/settingtypes.txt
Comment (Xeno333) [reply builtin/settingtypes.txt:268]: Ah yes... I will correct that

File: builtin/settingtypes.txt
Comment (grorp) [inline builtin/settingtypes.txt:269]: > Would you like me to make it "translucent_liquid" ?

I think this would be good:

```suggestion
#    Allows liquids to be translucent.
translucent_liquids (Translucent liquids) bool true
```

**Comment by Xeno333** [comment]
Just updated it to translucent_liquids from translucent_water for clearity

File: builtin/settingtypes.txt
Comment (Xeno333) [reply builtin/settingtypes.txt]: I just addedd that am and trynna squash commits rn, i have rly slow internet at the moment

File: builtin/settingtypes.txt
Comment (appgurueu) [reply builtin/settingtypes.txt]: (btw, for the future: you can resolve conversations once you have addressed the review points)

File: src/nodedef.cpp
Comment (grorp) [inline src/nodedef.cpp:289]:
```suggestion
	translucent_liquids            = g_settings->getBool("translucent_liquids");
	bool smooth_lighting           = g_settings->getBool("smooth_lighting");
	enable_mesh_cache              = g_settings->getBool("enable_mesh_cache");
	enable_minimap                 = g_settings->getBool("enable_minimap");
```

**Review by grorp — APPROVED** [review]

File: src/migratesettings.hpp
Comment (sfan5) [inline src/migratesettings.hpp:14]: We don't use the `.hpp` file extension and code should not be in a header file.

Comment (Xeno333) [reply src/migratesettings.hpp:14]: This is what  @Lars(appguru) came up with, as its not a standard header file. Its included that way because its only run one time. I advocated for doing it in a .cpp file and then having a header file but Lars thought it was to much for one function that will only be called once.

Comment (sfan5) [reply src/migratesettings.hpp:14]: I'll leave it to the other reviews to decide then.

Comment (grorp) [reply src/migratesettings.hpp:14]: > This is why we chose .hpp as to refrence it as a code header not a cpp file or regular header

doesn't make much sense to me, just didn't notice the different file extension

Comment (Xeno333) [reply src/migratesettings.hpp:14]: I can change it to .h but again its code too, i can just use .cpp and a sepreate .h but its only ever run one time. 


> > This is why we chose .hpp as to refrence it as a code header not a cpp file or regular header
> 
> doesn't make much sense to me, just didn't notice the different file extension


I can change to .h if you think its better

Comment (Xeno333) [reply src/migratesettings.hpp]: I renamed to .h insted of .hpp

Comment (appgurueu) [reply src/migratesettings.hpp]: > We don't use the `.hpp` file extension and code should not be in a header file.

This header is included exactly once, specifically in `main.cpp`, analogeous to `defaultsettings.h`, for a single function. There is no point in including it anywhere else. There will most certainly never be a point in including it in two different files.

`defaultsettings.cpp` has a corresponding header file, `defaultsettings.h`, which is also out of date. I do not see any benefit in the existence of `defaultsettings.h`.

It's just a function that's kept in a different file to not clutter `main.cpp`. As such, `.cpp` would be most appropriate, but including `.cpp` files is murky territory.

`.hpp` is the convention I use for "header-only" implementations such as this one. In our own sources, there is a weak precedent of this convention with `catch.hpp`.

I don't care about the file extension. But I do think that for this specific file there's no point in splitting it up in `.h` and `.cpp`.

Comment (Xeno333) [reply src/migratesettings.hpp]: > > We don't use the `.hpp` file extension and code should not be in a header file.
> 
> This header is included exactly once, specifically in `main.cpp`, analogeous to `defaultsettings.h`, for a single function. There is no point in including it anywhere else. There will most certainly never be a point in including it in two different files.
> 
> `defaultsettings.cpp` has a corresponding header file, `defaultsettings.h`, which is also out of date. I do not see any benefit in the existence of `defaultsettings.h`.
> 
> It's just a function that's kept in a different file to not clutter `main.cpp`. As such, `.cpp` would be most appropriate, but including `.cpp` files is murky territory.
> 
> `.hpp` is the convention I use for "header-only" implementations such as this one. In our own sources, there is a weak precedent of this convention with `catch.hpp`. This convention seems to be relatively popular among "header-only" C++ libraries more generally (for example `json.hpp`).
> 
> I don't care about the file extension. But I do think that for this specific file there's no point in splitting it up in `.h` and `.cpp`.

I agree, there is no point in splitting it when its only called by main. So I changed the file name to .h as the others do not seem to think using .hpp is good.

Comment (appgurueu) [reply src/migratesettings.hpp]: Seems fine to me. We have some header-only `.h` files already, though those are mostly templates, hence have no other choice.

I'll use `.h` for header-only files in Minetest in the future then.

**Comment by Xeno333** [comment]
for some reason the checks faild but i have like 100kbps internet rn so i cant check it

**Comment by appgurueu** [comment]
> for some reason the checks faild but i have like 100kbps internet rn so i cant check it

That's my fault. I have a fix, but I will wait for it to pass CI before pushing: https://github.com/minetest/minetest/compare/master...appgurueu:minetest:master.

File: src/migratesettings.h
Comment (appgurueu) [inline src/migratesettings.h:6]: Don't. [There is no point for this in C++.](https://softwareengineering.stackexchange.com/questions/286490/what-is-the-difference-between-function-and-functionvoid)

Comment (Xeno333) [reply src/migratesettings.h:6]: oh sorry I was just trying to get it to reload and it wouldnt without a change and i couldnt push bc i jave like 100kbps internet rn

Comment (Xeno333) [reply src/migratesettings.h]: oh sorry I was just trying to get it to reload and it wouldnt without a change and i couldnt push bc i jave like 100kbps internet rn

Comment (Xeno333) [reply src/migratesettings.h]: oh sorry I was just trying to get it to reload and it wouldnt without a change and i couldnt push bc i jave like 100kbps internet rn

**Comment by Xeno333** [comment]
> > for some reason the checks faild but i have like 100kbps internet rn so i cant check it
> 
> That's my fault. I have a fix, but I will wait for it to pass CI before pushing: [master...appgurueu:minetest:master](https://github.com/minetest/minetest/compare/master...appgurueu:minetest:master).

oh ok, its the reason my checks are faling?

**Comment by appgurueu** [comment]
> oh ok, its the reason my checks are faling?

Yes. I have restarted the failing check. It should pass now that I have pushed the commit.

**Comment by Xeno333** [comment]
> > oh ok, its the reason my checks are faling?
> 
> Yes. I have restarted the failing check. It should pass now that I have pushed the commit. Apologies for the inconvenience.

no problem

**Comment by Xeno333** [comment]
> > oh ok, its the reason my checks are faling?
> 
> Yes. I have restarted the failing check. It should pass now that I have pushed the commit. Apologies for the inconvenience.

ye it faild again... imma check logs my internet is a little faster now

**Comment by Xeno333** [comment]
your fork hasnt been merged...

**Pull request merged**
